devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy

```diff
@@ -14,736 +10,570 @@
+	log.debug "Configuring Reporting and Bindings ${device.name} ${device.deviceNetworkId} ${device.zigbeeId}"
+
+	//setup binding for 18 endpoints
+	def bindCmds = []
+	bindCmds += zigbee.addBinding(ON_OFF_CLUSTER, [destEndpoint: 1])
+	bindCmds += zigbee.addBinding(ALARMS_CLUSTER, [destEndpoint: 1])
+
+	for (endpoint in 1..18){
+		bindCmds += zigbee.addBinding(BINARY_INPUT_CLUSTER, [destEndpoint: endpoint])
+	}
+
+	//setup reporting for 18 endpoints
+	def reportingCmds = []
+	reportingCmds += zigbee.configureReporting(ON_OFF_CLUSTER, 0, DataType.BOOLEAN, 1, 0, 0x01, [destEndpoint: 1])
+	reportingCmds += zigbee.configureReporting(ALARMS_CLUSTER, 0, DataType.UINT16, 1, 0, 0x00, [destEndpoint: 1])
+
+	for (endpoint in 1..18){
+		reportingCmds += zigbee.configureReporting(BINARY_INPUT_CLUSTER, PRESENT_VALUE_IDENTIFIER, DataType.BOOLEAN, 1, 0, 0x01, [destEndpoint: endpoint])
+	}
+
+	return bindCmds + reportingCmds + setRainSensor()
 }
-def z13off() {
-	"st cmd 0x${device.deviceNetworkId} 14 6 0 {}"    
+
+//PING is used by Device-Watch in attempt to reach the Device
+def ping() {
+	if (DEBUG) log.debug "device health ping"
+	return zigbee.onOffRefresh()
 }
-def z14on() {
-	return manual() + "st cmd 0x${device.deviceNetworkId} 15 6 1 {}"
+
+def refresh() {
+	if (DEBUG) log.debug "refresh"
+
+	def refreshCmds = []
+	for (endpoint in 1..17){
+		refreshCmds += zigbee.readAttribute(BINARY_INPUT_CLUSTER, PRESENT_VALUE_IDENTIFIER, [destEndpoint: endpoint])
+	}
+	refreshCmds += zigbee.readAttribute(BINARY_INPUT_CLUSTER, OUT_OF_SERVICE_IDENTIFIER, [destEndpoint: 18])
+
+	return refreshCmds
 }
-def z14off() {
-    "st cmd 0x${device.deviceNetworkId} 15 6 0 {}"
+
+def healthPoll() {
+	if (DEBUG) log.debug "healthPoll()"
+	def cmds = refresh()
+	cmds.each { sendHubCommand(new physicalgraph.device.HubAction(it)) }
 }
-def z15on() {
-	return manual() + "st cmd 0x${device.deviceNetworkId} 16 6 1 {}"  
+
+def configureHealthCheck() {
+	Integer hcIntervalMinutes = 12
+	if (!state.hasConfiguredHealthCheck) {
+		log.debug "Configuring Health Check, Reporting"
+		unschedule("healthPoll")
+		runEvery5Minutes("healthPoll")
+		def healthEvent = [name: "checkInterval", value: hcIntervalMinutes * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID]]
+		// Device-Watch allows 2 check-in misses from device
+		sendEvent(healthEvent)
+		childDevices.each {
+			it.sendEvent(healthEvent)
+		}
+		state.hasConfiguredHealthCheck = true
+	}
 }
-def z15off() {
-	"st cmd 0x${device.deviceNetworkId} 16 6 0 {}"    
+
+//parse hex string and make integer
+private hextoint(String hex) {
+	Long.parseLong(hex, 16).toInteger()
 }
-def z16on() {
-	return manual() + "st cmd 0x${device.deviceNetworkId} 17 6 1 {}"    
+
+private hex(value) {
+	new BigInteger(Math.round(value).toString()).toString(16)
 }
-def z16off() {
-	"st cmd 0x${device.deviceNetworkId} 17 6 0 {}"    
+
+private String swapEndianHex(String hex) {
```

Comment:
You should be able to achieve what you want here using the builtins:

`zigbee.swapEndianHex(zigbee.convertToHexString(...))`